File: relay-node/src/protocol.js
/** Binary frame helpers — must match relay/app/protocol.py and shared/proto (Go). */

export const TYPE_HANDSHAKE = 0x01;
export const TYPE_HANDSHAKE_ACK = 0x02;
export const TYPE_RESUME = 0x03;
export const TYPE_FRAGMENT = 0x04;
export const TYPE_HANDSHAKE_AGG = 0xa1;

export const PROTO_VER = 1;
export const MTU_HINT = 1200;

export function parseResume(data) {
  if (data.length < 9 || data[0] !== TYPE_RESUME) return null;
  return Number(data.readBigUInt64BE(1));
}

export function parseHandshake(data) {
  if (data.length < 28 || data[0] !== TYPE_HANDSHAKE) return null;
  const shardId = data.readUInt16BE(2);
  const sessionId = Number(data.readBigUInt64BE(4));
  const nonce = data.subarray(12, 28);
  return { shardId, sessionId, nonce };
}

export function buildHandshakeAck(sessionId, relayShardId) {
  const out = Buffer.alloc(13);
  out[0] = TYPE_HANDSHAKE_ACK;
  out[1] = PROTO_VER;
  out.writeBigUInt64BE(BigInt(sessionId), 2);
  out.writeUInt16BE(MTU_HINT, 10);
  out[12] = relayShardId & 0xff;
  return out;
}

export function parseHandshakeAgg(data) {
  if (data.length < 10 || data[0] !== TYPE_HANDSHAKE_AGG) return null;
  const relayId = data[1];
  const sessionId = Number(data.readBigUInt64BE(2));
  return { relayId, sessionId };
}

export function isFragment(data) {
  return data.length >= 19 && data[0] === TYPE_FRAGMENT;
}

File: deploy/amvera/app/protocol.py
"""Binary frame helpers matching shared/proto (Go)."""

from __future__ import annotations

import struct

TYPE_HANDSHAKE = 0x01
TYPE_HANDSHAKE_ACK = 0x02
TYPE_RESUME = 0x03
TYPE_FRAGMENT = 0x04
TYPE_HANDSHAKE_AGG = 0xA1

PROTO_VER = 1
MTU_HINT = 1200


def parse_resume(data: bytes) -> int | None:
    if len(data) < 9 or data[0] != TYPE_RESUME:
        return None
    return struct.unpack_from(">Q", data, 1)[0]


def parse_handshake(data: bytes) -> tuple[int, int, bytes] | None:
    if len(data) < 28 or data[0] != TYPE_HANDSHAKE:
        return None
    shard_id = struct.unpack_from(">H", data, 2)[0]
    session_id = struct.unpack_from(">Q", data, 4)[0]
    nonce = data[12:28]
    return shard_id, session_id, nonce


def build_handshake_ack(session_id: int, relay_shard_id: int) -> bytes:
    return (
        bytes([TYPE_HANDSHAKE_ACK, PROTO_VER])
        + struct.pack(">QH", session_id, MTU_HINT)
        + bytes([relay_shard_id & 0xFF])
    )


def parse_handshake_agg(data: bytes) -> tuple[int, int] | None:
    if len(data) < 10 or data[0] != TYPE_HANDSHAKE_AGG:
        return None
    relay_id = data[1]
    session_id = struct.unpack_from(">Q", data, 2)[0]
    return relay_id, session_id


def is_fragment(data: bytes) -> bool:
    return len(data) >= 19 and data[0] == TYPE_FRAGMENT
